src/ParametersHandler/YarpImplementation/CMakeLists.txt

```diff
@@ -0,0 +1,57 @@
+# Copyright (C) 2020 Istituto Italiano di Tecnologia (IIT). All rights reserved.
+# This software may be modified and distributed under the terms of the
+# GNU Lesser General Public License v2.1 or any later version.
+
+# set target name
+if(BIPEDAL_LOCOMOTION_CONTROLLERS_COMPILE_YarpImplementation)
+
+  set(LIBRARY_TARGET_NAME ParametersHandlerYarpImplementation)
+
+  # set cpp files
+  set(${LIBRARY_TARGET_NAME}_SRC
+    src/YarpImplementation.cpp
+    )
+
+  # set hpp files
+  set(${LIBRARY_TARGET_NAME}_HDR
+    include/BipedalLocomotionControllers/ParametersHandler/YarpImplementation.h
+    include/BipedalLocomotionControllers/ParametersHandler/YarpImplementation.tpp
+    )
+
+  # add an executable to the project using the specified source files.
+  add_library(${LIBRARY_TARGET_NAME} SHARED ${${LIBRARY_TARGET_NAME}_SRC} ${${LIBRARY_TARGET_NAME}_HDR})
```

Comment:
When I try to compile `STATIC` I got the following error.
Probably there is something wrong with `YarpUtilities`

```
../../../../lib/libBipedalLocomotionControllersParametersHandlerYarpImplementation.a(YarpImplementation.cpp.o): In function `bool BipedalLocomotionControllers::YarpUtilities::getVectorFromSearchable<std::vector<bool, std::allocator<bool> > >(yarp::os::Searchable const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<bool, std::allocator<bool> >&)':
/home/gromualdi/robot-code/bipedal-locomotion-controllers/src/YarpUtilities/include/BipedalLocomotionControllers/YarpUtilities/Helper.tpp:168: multiple definition of `bool BipedalLocomotionControllers::YarpUtilities::getVectorFromSearchable<std::vector<bool, std::allocator<bool> > >(yarp::os::Searchable const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<bool, std::allocator<bool> >&)'
CMakeFiles/ParametesHandlerYarpTest.dir/ParametesHandlerYarpTest.cpp.o:/home/gromualdi/robot-code/bipedal-locomotion-controllers/src/YarpUtilities/include/BipedalLocomotionControllers/YarpUtilities/Helper.tpp:168: first defined here
collect2: error: ld returned 1 exit status
src/ParametersHandler/YarpImplementation/tests/CMakeFiles/ParametesHandlerYarpTest.dir/build.make:126: recipe for target 'bin/ParametesHandlerYarpTest' failed
make[2]: *** [bin/ParametesHandlerYarpTest] Error 1
CMakeFiles/Makefile2:1138: recipe for target 'src/ParametersHandler/YarpImplementation/tests/CMakeFiles/ParametesHandlerYarpTest.dir/all' failed
make[1]: *** [src/ParametersHandler/YarpImplementation/tests/CMakeFiles/ParametesHandlerYarpTest.dir/all] Error 2
Makefile:140: recipe for target 'all' failed
make: *** [all] Error 2
gromualdi@iiticublap109:~
```

Reply:
Probably it also needs to be compiled ``STATIC``. @traversaro once suggested me to avoid using ``SHARED`` when adding a library (simply put nothing). Then, in the main CMakeLists you add https://github.com/robotology/how-to-export-cpp-library/blob/master/CMakeLists.txt#L98

Reply:
When I try to follow what you suggested in https://github.com/dic-iit/bipedal-locomotion-controllers/pull/13#discussion_r376405015 I had the following error:
```
Scanning dependencies of target YarpUtilities
[  8%] Building CXX object src/YarpUtilities/CMakeFiles/YarpUtilities.dir/src/Helper.cpp.o
[ 16%] Linking CXX static library ../../lib/libBipedalLocomotionControllersYarpUtilities.a
[ 16%] Built target YarpUtilities
Scanning dependencies of target YarpUtilitiesTestMain
[ 25%] Building CXX object src/YarpUtilities/tests/CMakeFiles/YarpUtilitiesTestMain.dir/YarpUtilitiesTestMain.cpp.o
[ 33%] Linking CXX static library ../../../lib/libYarpUtilitiesTestMain.a
[ 33%] Built target YarpUtilitiesTestMain
[ 41%] Linking CXX executable ../../../bin/YarpUtilitiesTest
../../../lib/libBipedalLocomotionControllersYarpUtilities.a(Helper.cpp.o): In function `bool BipedalLocomotionControllers::YarpUtilities::getVectorFromSearchable<std::vector<bool, std::allocator<bool> > >(yarp::os::Searchable const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<bool, std::allocator<bool> >&)':
/home/gromualdi/robot-code/bipedal-locomotion-controllers/src/YarpUtilities/include/BipedalLocomotionControllers/YarpUtilities/Helper.tpp:168: multiple definition of `bool BipedalLocomotionControllers::YarpUtilities::getVectorFromSearchable<std::vector<bool, std::allocator<bool> > >(yarp::os::Searchable const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<bool, std::allocator<bool> >&)'
CMakeFiles/YarpUtilitiesTest.dir/YarpUtilitiesTest.cpp.o:/home/gromualdi/robot-code/bipedal-locomotion-controllers/src/YarpUtilities/include/BipedalLocomotionControllers/YarpUtilities/Helper.tpp:168: first defined here
collect2: error: ld returned 1 exit status
src/YarpUtilities/tests/CMakeFiles/YarpUtilitiesTest.dir/build.make:125: recipe for target 'bin/YarpUtilitiesTest' failed
make[2]: *** [bin/YarpUtilitiesTest] Error 1
CMakeFiles/Makefile2:1055: recipe for target 'src/YarpUtilities/tests/CMakeFiles/YarpUtilitiesTest.dir/all' failed
make[1]: *** [src/YarpUtilities/tests/CMakeFiles/YarpUtilitiesTest.dir/all] Error 2
Makefile:140: recipe for target 'all' failed
```

I should open a issue

Reply:
Yes, you can avoid facing this now.

Reply:
Ah, some good old One Definition Rule (ODR, https://en.wikipedia.org/wiki/One_Definition_Rule) violations. I do  not know the details, but the tipical trick is to mark  the offending function as `inline` or avoid explicit instantiation.

Reply:
Other useful references: 
* https://en.cppreference.com/w/cpp/language/definition
* https://gieseanw.wordpress.com/2018/10/30/oops-i-violated-odr-again/